fdbserver/CommitProxyServer.actor.cpp

```diff
@@ -874,6 +876,104 @@ ACTOR Future<Void> preresolutionProcessing(CommitBatchContext* self) {
 	return Void();
 }
 
+namespace {
+ErrorOr<Optional<std::pair<TenantName, int64_t>>> getTenantDetails(ProxyCommitData* commitData,
+                                                                   Optional<int64_t> tenantId,
+                                                                   bool logOnFailure) {
+	if (tenantId.present() && tenantId.get() != TenantInfo::INVALID_TENANT) {
+		auto itr = commitData->tenantIdNameMap.find(tenantId.get());
+		if (itr == commitData->tenantIdNameMap.end()) {
+			if (logOnFailure) {
+				TraceEvent(SevWarn, "CommitProxyUnknownTenant", commitData->dbgid).detail("Tenant", tenantId.get());
+			}
+
+			return unknown_tenant();
+		}
+
+		return ErrorOr<Optional<std::pair<TenantName, int64_t>>>(std::make_pair(itr->second, itr->first));
+	}
+
+	return Optional<std::pair<TenantName, int64_t>>();
+}
+
+std::pair<EncryptCipherDomainName, int64_t> getEncryptDetailsFromMutationRef(ProxyCommitData* commitData,
+                                                                             MutationRef m) {
+	std::pair<EncryptCipherDomainName, int64_t> details(EncryptCipherDomainName(), TenantInfo::INVALID_TENANT);
+
+	// Possible scenarios:
+	// 1. Encryption domain (Tenant details) weren't explicitly provided, extract Tenant details using
+	// TenantPrefix (first 8 bytes of FDBKey)
+	// 2. Encryption domain isn't available, leverage 'default encrypiton domain'
+	bool useDefaultEncryptionDomain = true;
+	if (isSingleKeyMutation((MutationRef::Type)m.type)) {
+		if (isSystemKey(m.param1)) {
+			// Encryption domain == FDB SystemKeyspace encryption domain
+			details.first = EncryptCipherDomainName(FDB_DEFAULT_ENCRYPT_DOMAIN_NAME);
+			details.second = SYSTEM_KEYSPACE_ENCRYPT_DOMAIN_ID;
+		} else if ((MutationRef::Type)m.type == MutationRef::Type::ClearRange) {
+			// FIXME: Handle Clear-range transaction, actions needed:
+			// 1. Transaction range can spawn multiple encryption domains (tenants)
+			// 2. Transaction can be a multi-key transaction spawning multiple tenants
+			// For now fallback to 'default encryption domain'
+			ASSERT(useDefaultEncryptionDomain);
+
+			CODE_PROBE(true, "SingleKey ClearRange mutation encryption");
+		} else if (m.param1.size() >= TENANT_PREFIX_SIZE) {
```

Comment:
Wondering if a key that's exactly the tenant prefix belong to the tenant cc @sfc-gh-ajbeamon

Reply:
Yes, a tenant can contain the empty key, in which case it would only have the tenant prefix.